File: 0200-number-of-islands/0200-number-of-islands.js
var numIslands = function(grid) {
    if (!grid || grid.length === 0) return 0;

    const directions = [
        [0, 1],  // 오른쪽
        [0, -1], // 왼쪽
        [1, 0],  // 아래
        [-1, 0]  // 위
    ];

    const m = grid.length;
    const n = grid[0].length;
    let number = 0;

    const bfs = (x, y) => {
        const queue = [[x, y]];
        grid[x][y] = "0";  // 방문 처리

        while (queue.length > 0) {
            const [curX, curY] = queue.shift();

            for (const [dx, dy] of directions) {
                const nx = curX + dx;
                const ny = curY + dy;

                if (nx >= 0 && nx < m && ny >= 0 && ny < n && grid[nx][ny] === "1") {
                    queue.push([nx, ny]);
                    grid[nx][ny] = "0"; // 방문 처리
                }
            }
        }
    };

    for (let i = 0; i < m; i++) {
        for (let j = 0; j < n; j++) {
            if (grid[i][j] === "1") {
                bfs(i, j);
                number++;
            }
        }
    }

    return number;
};

File: 0200-number-of-islands/0200-number-of-islands.py
from collections import deque
from typing import List

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or not grid[0]:
            return 0

        rows, cols = len(grid), len(grid[0])
        islands = 0

        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]

        def bfs(x: int, y: int):
            queue = deque()
            queue.append((x, y))
            grid[x][y] = '0'  # 방문 처리

            while queue:
                cx, cy = queue.popleft()
                for dx, dy in directions:
                    nx, ny = cx + dx, cy + dy
                    if 0 <= nx < rows and 0 <= ny < cols and grid[nx][ny] == '1':
                        queue.append((nx, ny))
                        grid[nx][ny] = '0'

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == '1':
                    bfs(i, j)
                    islands += 1

        return islands